In [12]:
import pandas as pd
import numpy as np
from sklearn.decomposition import FactorAnalysis
from factor_analyzer.factor_analyzer import calculate_kmo
from sklearn.preprocessing import StandardScaler

# Step 1: Load the CSV file
data = pd.read_csv('kc_house_data_reduced.csv')

# Step 2: Extract relevant features for factor analysis
features = data[['condition', 'grade', 'sqft_above', 'sqft_basement', 'sqft_living15']]

# Step 3: Standardize the features (important for factor analysis)
scaler = StandardScaler()
features_scaled = scaler.fit_transform(features)

# Step 4: Fit a factor analysis model with two components
n_components = 2
factor_model = FactorAnalysis(n_components=n_components, random_state=42)
factor_model.fit(features_scaled)


# Step 5: Print out the components found through factor analysis
components = factor_model.components_
print("Factor Analysis Components:")
print(components)

# Step 6: Interpret the components in plain English
component_names = ['Quality', 'Size']

for i, component in enumerate(components):
    print(f"\nInterpretation for Component {i + 1} ({component_names[i]}):")
    
    # Sort the indices by the absolute values of the component weights
    sorted_indices = abs(component).argsort()[::-1]
    
    # Print the top features and their weights in the component
    for j in sorted_indices:
        print(f"{features.columns[j]}: {component[j]:.3f}")
# Step 7: Calculate KMO using factor_analyzer
print("  ")
kmo_all, kmo_model = calculate_kmo(features_scaled)

print(f"Overall KMO: {kmo_all}")
print(f"KMO for the model: {kmo_model}")

Factor Analysis Components:
[[-0.06915843  0.82146164  0.78397372  0.44373017  0.81491862]
 [ 0.24241864 -0.23514799 -0.47719398  0.83719893 -0.19252606]]

Interpretation for Component 1 (Quality):
grade: 0.821
sqft_living15: 0.815
sqft_above: 0.784
sqft_basement: 0.444
condition: -0.069

Interpretation for Component 2 (Size):
sqft_basement: 0.837
sqft_above: -0.477
condition: 0.242
grade: -0.235
sqft_living15: -0.193
  
Overall KMO: [0.66328507 0.73314139 0.6464693  0.26537095 0.74692203]
KMO for the model: 0.6724465791116416


The Factor Analysis Components reveal two latent variables: Quality and Size. Quality is primarily represented by high positive loadings on grade (0.821), sqft_living15 (0.815), sqft_above (0.784), and sqft_basement (0.444), indicating that houses with superior construction and condition contribute positively. Size, on the other hand, is characterized by positive loadings on sqft_basement (0.837) and condition (0.242), but negative loadings on sqft_above (-0.477), grade (-0.235), and sqft_living15 (-0.193), suggesting that larger size, along with better condition and grade, influences this dimension.

The Overall Kaiser-Meyer-Olkin (KMO) measure of sampling adequacy is 0.672, indicating moderate suitability for factor analysis. However, individual KMO values for each variable vary, suggesting that while the model is reasonably valid, there may be some variables less suited for inclusion.The factors are interpretable, aligning with my colleague's suggested model, capturing non-price aspects effectively. Any lack of interpretability may stem from variable complexities or correlations not fully captured by the selected components.In simpler terms, while the identified factors make sense in the context of quality and size, there might be additional intricacies in the data that the two selected components may not entirely capture or explain.